src/cargo/ops/cargo_package.rs

```diff
@@ -178,34 +250,100 @@ pub fn package(ws: &Workspace<'_>, opts: &PackageOpts<'_>) -> CargoResult<Option
         // below, and will be validated during the verification step.
     }
 
-    for (pkg, cli_features) in pkgs {
+    let just_pkgs: Vec<_> = pkgs.iter().map(|p| p.0).collect();
+    let publish_reg = infer_registry(ws.gctx(), &just_pkgs, opts.reg_or_index.clone())?;
+    debug!("packaging for registry {publish_reg}");
+
+    let reg_dir = ws.target_dir().join("package").join("tmp-registry");
+    let mut local_reg = TmpRegistry::new(ws.gctx(), reg_dir, publish_reg)?;
+
+    let deps = local_deps(ws, pkgs.iter().map(|(p, f)| ((*p).clone(), f.clone())));
+
+    // Packages need to be created in dependency order, because dependencies must
+    // be added to our local overlay before we can create lockfiles that depend on them.
+    let sorted_pkgs = deps.sort();
+    let mut outputs: Vec<(Package, PackageOpts<'_>, FileLock)> = Vec::new();
+    for (pkg, cli_features) in sorted_pkgs {
         let opts = PackageOpts {
+            cli_features: cli_features.clone(),
             to_package: ops::Packages::Default,
-            cli_features,
             ..opts.clone()
         };
-        let ar_files = build_ar_list(ws, pkg, &opts)?;
+        let ar_files = build_ar_list(ws, &pkg, &opts)?;
 
         if opts.list {
-            for ar_file in ar_files {
+            for ar_file in &ar_files {
                 drop_println!(ws.gctx(), "{}", ar_file.rel_str);
             }
         } else {
-            let tarball = create_package(ws, pkg, ar_files)?;
-            if opts.verify {
-                run_verify(ws, pkg, &tarball, &opts)
-                    .with_context(|| "failed to verify package tarball")?;
-            }
-            dsts.push(tarball);
+            let tarball = create_package(ws, &pkg, ar_files, Some(&local_reg))?;
+            local_reg.add_package(&pkg, &tarball)?;
+            outputs.push((pkg, opts, tarball));
         }
     }
 
-    if opts.list {
-        // We're just listing, so there's no file output
-        Ok(None)
-    } else {
-        Ok(Some(dsts))
+    // Verify all packages in the workspace. This can be done in any order, since the dependencies
+    // are already all in the local registry overlay.
+    if opts.verify {
```

Comment:
Are we no verifying when using `--list`?

Reply:
No, and this matches the old behavior, where with `--list` we return early before getting to `run_verify`.